packages/shylu/shylu_dd/frosch/src/CoarseSpaces/FROSch_CoarseSpace_def.hpp

```diff
@@ -51,77 +51,127 @@ namespace FROSch {
+        FROSCH_ASSERT(AssembledBasisMap_->getMinAllGlobalIndex()==AssembledBasisMapUnique_->getMinAllGlobalIndex(),"FROSch::CoarseSpace : ERROR: AssembledBasisMap_->getMinAllGlobalIndex()!=AssembledBasisMapUnique_->getMinAllGlobalIndex()");
+        FROSCH_ASSERT(GO(AssembledBasisMapUnique_->getGlobalNumElements())==GO(AssembledBasisMapUnique_->getMaxAllGlobalIndex()+1),"FROSch::CoarseSpace : ERROR: AssembledBasisMapUnique_->getGlobalNumElements()!=(AssembledBasisMapUnique_->getMaxAllGlobalIndex()+1)");
+        
+        // Basis
+        if (!AssembledBasisMap_.is_null()) {
             if (AssembledBasisMap_->getGlobalNumElements()>0) { // AH 02/12/2019: Is this the right condition? Seems to work for now...
-                AssembledBasis_ = MultiVectorFactory<SC,LO,GO,NO >::Build(SerialRowMap_,AssembledBasisMap_->getNodeNumElements());
-                for (UN i=0; i<UnassembledBasesMaps_.size(); i++) {
-                    for (UN j=0; j<UnassembledBasesMaps_[i]->getNodeNumElements(); j++) {
-                        AssembledBasis_->getDataNonConst(itmp).deepCopy(UnassembledSubspaceBases_[i]->getData(j)()); // Here, we copy data. Do we need to do this?
-                        itmp++;
+                LO totalSize = -1;
+                for (UN i=0; i<UnassembledSubspaceBases_.size(); i++) {
+                    if (!UnassembledSubspaceBases_[i].is_null()) totalSize = std::max(totalSize,LO(UnassembledSubspaceBases_[i]->getLocalLength()+Offsets_[i]));
+                }
+                XMapPtr serialMap = MapFactory<LO,GO,NO>::Build(AssembledBasisMap_->lib(),totalSize,0,this->SerialComm_);
+                
+                AssembledBasis_ = MultiVectorFactory<SC,LO,GO,NO >::Build(serialMap,AssembledBasisMap_->getNodeNumElements());
+                for (UN i=0; i<UnassembledSubspaceBases_.size(); i++) {
+                    if (!UnassembledSubspaceBases_[i].is_null()) {
+                        for (UN j=0; j<UnassembledSubspaceBases_[i]->getNumVectors(); j++) {
+                            for (UN k=0; k<UnassembledSubspaceBases_[i]->getLocalLength(); k++) {
+                                FROSCH_ASSERT(itmp<AssembledBasis_->getNumVectors(),"FROSch::CoarseSpace : ERROR: itmp>=AssembledBasis_->getNumVectors()");
+                                FROSCH_ASSERT(k+Offsets_[i]<AssembledBasis_->getLocalLength(),"FROSch::CoarseSpace : ERROR: k+Offsets_[i]>=AssembledBasis_->getLocalLength()");
+                                AssembledBasis_->replaceLocalValue(k+Offsets_[i],itmp,UnassembledSubspaceBases_[i]->getData(j)[k]);
+                            }
+                            itmp++;
+                        }
                     }
                 }
             }
         }
 
         UnassembledBasesMaps_.resize(0);
```

Comment:
Resizing a `std::vector` (or `Teuchos::Array`) does not necessarily clear its memory.  This means that the `RCP<const Map>` objects might not get freed here.  The standard idiom for clearing a `std::vector` is to swap it with an empty vector, like this:
```c++
std::vector<whatever> new_vector;
std::swap(old_vector, new_vector);
```

Reply:
Those are `Teuchos::Array`s. Could I also use `clear()`?

Reply:
`Teuchos::Array` is just a wrapper around `std::vector`.  It does what `std::vector` (as of C++98) does.